Add KEP for application pod graceful update

zhan849 opened the pull request:
This is the initial draft of KEP addressing issue https://github.com/kubernetes/kubernetes/issues/75136

/sig apps
/area controller
/assign @kow3ns 
/cc @janetkuo

zhan849 commented:
I signed it

janetkuo commented on keps/sig-apps/0033-20190316-app-pods-graceful-update.md:
```diff
@@ -0,0 +1,191 @@
+    // PatchIfPossiblePodUpdateStrategyType indicates that we try to patch Pod instead of
+    // recreate Pod when possible. Currently we patch Pod only when any of the main
+    // containers (those in Spec.Containers) has image changes
+    PatchIfPossiblePodUpdateStrategyType = "PatchIfPossible"
```
It's unclear from the comment whether a pod will be recreated when fields other than container image get updated. 

The implementation details should also cover how controller revision history is handled during a rollout, and how to compare pod spec against revision history or pod template to know whether fields other than image is updated.

zhan849 replied:
Thanks for your suggestion. Will fix the comment and add details about revision history handling and implementations about checking whether pod should be patched or recreated

FillZpp replied:
Maybe "InPlaceIfPossible" ?

zhan849 replied:
Thanks for the suggestions. I'd vote on `InPlaceIfPossible`, as for now we only upgrade in place when container image changes.

janetkuo commented:
/ok-to-test

FillZpp commented on keps/sig-apps/0033-20190316-app-pods-graceful-update.md:
```diff
@@ -0,0 +1,191 @@
+
+    // RecreatePodUpdateStrategyType indicates that we always delete Pod and create new Pod
+    // during Pod update, which is the current behavior
+    RecreatePodUpdateStrategyType = "ReCreate"
```
I think "ReCreate" here is more appropriate than "Delete". 
On the other hand, "Delete" might be confusing because of the "OnDelete" StatefulSetUpdateStrategyType.

zhan849 replied:
Yep "DeletePodUpdateStrategyType" also sounds weird therefore I'm more inclined to "ReCreate". Currently we have no plan to support pod in place upgrade for deployment as it's upgrade mechanism is a bit different by using replica set.

Event we want to support this in deployment, we can put this in to `RollingUpdateDeployment` so it will only appear in rolling update and thus less confusion.

I can change the name from "ReCreate" to "Recreate" - just we are already using this as one word in k8s

draveness commented:
Hi @zhan849, I'm very interested in this proposal. How is this going on?

zhan849 commented:
Hi @draveness sorry about the delay as I just went through a crazy Q2 and have this KEP lagged behind. My eta is to finalize addressing several concerns pointed out here and have it updated for another round of review by end of next week.

zhan849 commented:
@janetkuo sry about the late update - I addressed your comments, added sections about how to determine whether the change is eligible for in place update, and more details about revision history handling / component change. Please check.

gaorong commented:
thanks for this awesome feature, truly helpful. it also a demand for `deployment/replicaset`.  could you please also add this to deployment.

zhan849 commented:
> thanks for this awesome feature, truly helpful. it also a demand for `deployment/replicaset`. could you please also add this to deployment.

@gaorong thanks for the feed back. I've thought about adding in place upgrade support for deployment / replicaset but finally decided to be mover conservative on this and leave inplace upgrade for deployment / replica set as non-goal.

The reason is that, unlike statefulset/daemonset, whose replicas are unique and delete/recreate kind of upgrade could **affect availability**, deployment/replicaset are supposed to be stateless, and current upgrade strategy is creating new replica first, which **does not** affect availability like the former two. Not sure if in-place update is on the same path of the design principal of "stateless" application in k8s. @janetkuo or @kow3ns can you guys provide thoughts?

also, @gaorong can you pls provide more details about the painpoints you are having? thanks!

gaorong commented:
Although deployment is not as painful as statefulset, but It's still painful more or less. Why don't we implement this by the way?

skilxn-go commented:
@zhan849 Hi, in our company and use case, we are strongly depend on this feature. As a matter of fact, we already implement this by a CRD and a custom controller. So I don't know if it's ok for us to working on this KEP together. As I see, may be you don't have enough time to follow this KEP very timely. If it's ok, please give me your contact details, and let's work on this.
